cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -0,0 +1,427 @@
+  auto buf = Buffer::FromString(bytes);
+  auto type = fixed_size_binary(static_cast<int>(buf->size()));
+  return std::make_shared<FixedSizeBinaryScalar>(std::move(buf), std::move(type));
+}
+
+}  // namespace
+
+Result<compute::Expression> FromProto(const st::Expression& expr) {
+  switch (expr.rex_type_case()) {
+    case st::Expression::kLiteral: {
+      ARROW_ASSIGN_OR_RAISE(auto datum, FromProto(expr.literal()));
+      return compute::literal(std::move(datum));
+    }
+
+    default:
+      break;
+  }
+
+  return Status::NotImplemented("conversion to arrow::compute::Expression from ",
+                                expr.DebugString());
+}
+
+Result<Datum> FromProto(const st::Expression::Literal& lit) {
+  switch (lit.literal_type_case()) {
+    case st::Expression::Literal::kBoolean:
+      return Datum(lit.boolean());
+
+    case st::Expression::Literal::kI8:
+      return Datum(static_cast<int8_t>(lit.i8()));
+    case st::Expression::Literal::kI16:
+      return Datum(static_cast<int16_t>(lit.i16()));
+    case st::Expression::Literal::kI32:
+      return Datum(static_cast<int32_t>(lit.i32()));
+    case st::Expression::Literal::kI64:
+      return Datum(static_cast<int64_t>(lit.i64()));
+
+    case st::Expression::Literal::kFp32:
+      return Datum(lit.fp32());
+    case st::Expression::Literal::kFp64:
+      return Datum(lit.fp64());
+
+    case st::Expression::Literal::kString:
+      return Datum(lit.string());
+    case st::Expression::Literal::kBinary:
+      return Datum(std::make_shared<BinaryScalar>(Buffer::FromString(lit.binary())));
+
+    case st::Expression::Literal::kTimestamp:
+      return Datum(std::make_shared<TimestampScalar>(
+          static_cast<int64_t>(lit.timestamp()), TimeUnit::MICRO));
+
+    case st::Expression::Literal::kTimestampTz:
+      return Datum(std::make_shared<TimestampScalar>(
+          static_cast<int64_t>(lit.timestamp_tz()), TimeUnit::MICRO,
+          TimestampTzTimezoneString()));
+
+    case st::Expression::Literal::kDate:
+      return Datum(std::make_shared<Date32Scalar>(static_cast<int32_t>(lit.date())));
+    case st::Expression::Literal::kTime:
+      return Datum(std::make_shared<Time64Scalar>(static_cast<int64_t>(lit.time()),
+                                                  TimeUnit::MICRO));
+
+    case st::Expression::Literal::kIntervalYearToMonth:
+    case st::Expression::Literal::kIntervalDayToSecond:
+      break;
+
+    case st::Expression::Literal::kUuid:
+      return Datum(std::make_shared<ExtensionScalar>(
+          FixedSizeBinaryScalarFromBytes(lit.uuid()), uuid()));
+
+    case st::Expression::Literal::kFixedChar:
+      return Datum(std::make_shared<ExtensionScalar>(
+          FixedSizeBinaryScalarFromBytes(lit.fixed_char()),
+          fixed_char(static_cast<int32_t>(lit.fixed_char().size()))));
+
+    case st::Expression::Literal::kVarChar:
+      // FIXME
+      // There's no way to determine VarChar.length from the literal
```

Comment:
I don't think this matters, since the length is an upper bound. There's no real benefit to returning something other than plain old string here for now, because I don't think Arrow is currently capable of representing a limited-variable-length string right now, and I personally do not think that is important to address any time soon. `string` can represent this just fine.